### DISCUSSION TITLE
Slug - any package ?

### DISCUSSION THREAD

**Discussion by dawidow** [adonisjs/core#1814] [Help] [answered]
Hello. Is there any official/unofficial package that helps to create **slug** in our Database. I'm creating a blog and want to display instead of **id** -> **slug**. I don't know how to get down to it. There are a lot of helpers in laravel. How does it look in AdonisJS?

**Comment by thetutlage** [comment, marked as answer]
It depends what you are looking for. If its just creating slugs, then you can use https://www.npmjs.com/package/@slynova/slug package and make use of the `beforeSave` hook to create and update the slug as follows.

```ts
class Post extends BaseModel {
  @beforeSave()
  public static generateSlug(post: Post) {
    if (post.$dirty.title) {
      post.slug = slugify(post.title)
    }
  }
}
```

Next, you must add a unique constraint on the `slug` column to avoid duplicate slugs. It means if two posts have the same title, then you will end up with an exception raised by the database.

And to ensure uniqueness on posts with same title, you will have to do bunch of extra work and there was a package in v4 to handle this use case https://github.com/adonisjs/adonis-lucid-slugify

I will be migrating this package for v5 soon. However, for a simple tutorial, the above share technique should do the job

**Comment by mxthevs** [comment]
Here is what I'm doing to generate unique slugs while `adonis-lucid-slugify` doesn't get migrated to V5. It's a very simple version of the package, but it works.

```ts
import { LucidModel } from '@ioc:Adonis/Lucid/Model'
import slugify from '@slynova/slug'

export default class Slug {
  public static async make<T extends LucidModel>(
    modelInstance: T,
    fieldValue: string
  ): Promise<string> {
    const slug: string = slugify(fieldValue)

    const alreadyExists = await modelInstance
      .query()
      .select('*')
      .whereRaw(`?? ~* ?`, ['slug', `^${slug}(-[0-9]*)?$`]) // PostgreSQL only
      .orderBy('created_at', 'desc')
      .first()

    if (!alreadyExists) {
      return slug
    }

    const lastNum = Number(alreadyExists['slug'].replace(`${slug}-`, ''))
    return !lastNum || Number.isNaN(lastNum) ? `${slug}-1` : `${slug}-${lastNum + 1}`
  }
}
```
```ts
@beforeSave()
  public static async generateSlug(customer: Customer) {
    if (customer.$dirty.name) {
      customer.slug = await Slug.make(this, customer.name)
    }
  }
```

**Reply by MriDx** [reply]
for mysql --
```
.whereRaw(`?? REGEXP ?`, ['slug', `^${slug}(-[0-9]*)?$`]) // MySQL
```
